### PR TITLE
Merge issue43 hotfix

### LINKED ISSUE #43: Watcher::store_triggered_appointment should be called as an async task
The current codebase has an edge case where, if `bitcoind` is unreachable while an already triggered appointment is added, and `store_triggered_appointment` is called as result, the request will hang and end up timing out but be accepted. However, the user will not be handed the receipt.

This could be fixed making `store_triggered_appointment` be called as an async task, so the receipt can be handed straightaway. This is surprisingly not as straightforward as it sounds, and implies dealing with lifetimes in a way I don't fully understand (at least with my current knowledge of Rust 😅).

### PR DESCRIPTION
**Pull request by carterian8**
Closes #43.  `Watcher::add_appointment` spawns an `async` `tokio::task` to handle cases when `Watcher::store_triggered_appointment` runs into `bitcoind` connectivity issues.

Introducing this `async` change caused a substantial ripple effect through the code with the most notable being an update to the `Carrier` to use `tokio::sync::Notify` to listen for changes to the status of `bitcoind_reachable` (please see #43 for justification).  Other changes just further support the constraints introduced by incorporating more `async` operations.

All unit tests were updated to verify these changes and an additional test, `Watcher::test_add_appointment_bitcoind_unreachable`, was added to verify the correctness of the specific corner case presented in #43.

### CODE REVIEW

File: teos/src/carrier.rs [hunk 1]
```diff
@@ -72,153 +79,200 @@ impl Carrier {
...
-        log::info!("Pushing transaction to the network: {}", tx.txid());
-        let receipt = match self.bitcoin_cli.send_raw_transaction(tx) {
-            Ok(_) => {
-                // Here the transaction could, potentially, have been in mempool before the current height.
-                // This shouldn't really matter though.
-                log::info!("Transaction successfully delivered: {}", tx.txid());
-                ConfirmationStatus::InMempoolSince(self.block_height)
+    pub(crate) async fn send_transaction(&self, tx: &Transaction) -> ConfirmationStatus {
+        let mut continue_looping = true;
+        let mut receipt: Option<ConfirmationStatus> = None;
```
Comment (sr-gi) [inline teos/src/carrier.rs:84]: This shouldn't be needed. I think it comes from the `continue_looping` approach so in the case you need to loop again you have no value for `receipt`. I think recursion may make more sense.

**Review by sr-gi — COMMENTED** [review]
I gave this a first pass. It really feels like this should be solvable in an easier way.

Going the async route means we need to access async functions from `chain::Listen` (as you already mentioned in #43) which means the design has to be changed significantly.  On top of that, it also means accessing the Carrier in an async way, but this interacts with bitcoind synchronously, so things become pretty hacky.

I need deeper understanding on Rust async and lifetimes to judge how to properly fix this. Given this fixes a pretty rare edge case I'll leave it on hold for now.

PS: I've created a branch trying to simplify this but didn't really go further before realizing what I stated in the review https://github.com/talaia-labs/rust-teos/tree/55-tokio-notify. I will revisit this once the cln plugin is merged.

**Comment by carterian8** [comment]
@sr-gi

Yeah I agree this fix is a bit meaty for the problem it addresses.

Regarding lifetimes and Rust, I'm not sure you'll find a clean solution combining `async` and `self`.  I stumbled across some workarounds but, based on my understanding of Rust, referring to `self` within an `async` call violates the `static` lifetime requirement and calls for some unconventional techniques.

In my opinion, if you'd like to use `async` to fix this issue, its going to require a redesign of the magnitude presented here.

Maybe we can chat when you're ready and converge on a solution you'd be happy with.

File: teos/src/carrier.rs [hunk 2]
```diff
@@ -72,153 +79,200 @@ impl Carrier {
...
+        while continue_looping {
```
[the remark this answers is not in the document]
Comment (carterian8) [reply teos/src/carrier.rs:85]: Certainly open to `async_recursion` here but usually tend towards non-recursive solutions to save the stack.

File: teos/src/watcher.rs [hunk 3]
```diff
@@ -439,12 +478,13 @@ impl Watcher {
     ///
     /// If the appointment is rejected by the [Responder] (i.e. for being invalid), the data is wiped
     /// from the database but the slot is not freed.
-    fn store_triggered_appointment(
-        &self,
```
[the remark this answers is not in the document]
Comment (carterian8) [reply teos/src/watcher.rs:443]: Yeah this didn't feel right to me either but based on my understanding of Rust this is inevitable because you cannot use `self` within an `async` call as it violates the `static` lifetime requirement.

File: teos/src/carrier.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (carterian8) [reply teos/src/carrier.rs:85]: Again this is just a side effect of the `async` change made in `Watcher`

File: teos/src/watcher.rs [hunk 4]
```diff
@@ -806,50 +860,87 @@ impl chain::Listen for Watcher {
...
-                );
+            let responder = self.responder.clone();
+            let appointments = self.appointments.clone();
+            let locator_uuid_map = self.locator_uuid_map.clone();
+            let gatekeeper = self.gatekeeper.clone();
+            let dbm = self.dbm.clone();
+            join_handle_option = Some(tokio::spawn(async move {
+                let mut appointments_to_delete = HashSet::from_iter(
+                    invalid_breaches.into_keys());
+                let mut delivered_appointments = HashSet::new();
```
Comment (mariocynicys) [inline teos/src/watcher.rs:871]: Note: This would make this method exit early and a task will be spawned on tokio to do the job. This could potentially make the responder's/gatekeeper's own `block_connected` method run before or interleaving with this task.
